end2end-tests/src/test/kotlin/com/cjbooms/fabrikt/clients/okio3/Okio3Test.kt

```diff
@@ -0,0 +1,224 @@
+
+        assertThat(result.data).isEqualTo(
+            QueryResult(
+                listOf(FirstModel(id = testInfo.displayName))
+            )
+        )
+    }
+
+    @Test
+    fun `adds additional headers to the query`(testInfo: TestInfo) {
+        wiremock.get {
+            urlPath like "/example-path-1"
+            headers contains "awesome" like "header"
+        } returns {
+            statusCode = 200
+            body = mapper.writeValueAsString(
+                QueryResult(
+                    listOf(FirstModel(id = testInfo.displayName))
+                )
+            )
+        }
+
+        val result = examplePath1Client.getExamplePath1(additionalHeaders = mapOf("awesome" to "header"))
+
+        assertThat(result.data).isEqualTo(
+            QueryResult(
+                listOf(FirstModel(id = testInfo.displayName))
+            )
+        )
+    }
+
+    @Test
+    fun `send body with post request`(testInfo: TestInfo) {
+        val content = FirstModel(id = testInfo.displayName)
+        wiremock.post {
+            urlPath like "/example-path-1"
+            body equalTo mapper.writeValueAsString(content)
+        } returns {
+            statusCode = 201
+        }
+
+        val result = examplePath1Client.postExamplePath1(content)
+        assertThat(result.statusCode).isEqualTo(201)
+    }
+
+    @ParameterizedTest
+    @MethodSource("path2ErrorCodes")
+    fun `throws an exception if a 4xx http status code is returned`(errorCode: Int) {
+        wiremock.get {
+            urlPath like "/example-path-2/$errorCode"
+        } returns {
+            statusCode = errorCode
+            body = mapper.writeValueAsString(failure)
+        }
+
+        val result = assertThrows<ApiClientException> {
+            examplePath2Client.getExamplePath2PathParam(errorCode.toString(), 10)
+        }
+
+        assertThat(result.statusCode).isEqualTo(errorCode)
+        assertThat(mapper.readValue(result.message, Failure::class.java)).isEqualTo(failure)
+    }
+
+    @Test
+    fun `throws an exception if a http status code 500 is returned`() {
+        wiremock.get {
+            urlPath like "/example-path-2/500"
+        } returns {
+            statusCode = 500
+            body = mapper.writeValueAsString(failure)
+        }
+
+        val result = assertThrows<ApiServerException> {
+            examplePath2Client.getExamplePath2PathParam("500", 10)
+        }
+
+        assertThat(result.statusCode).isEqualTo(500)
+        assertThat(mapper.readValue(result.message, Failure::class.java)).isEqualTo(failure)
+    }
+
+    @Test
+    fun `throws an exception if a http status code 304 is returned`() {
+        wiremock.get {
+            urlPath like "/example-path-2/304"
+        } returns {
+            statusCode = 304
+        }
+
+        val result = assertThrows<ApiException> {
+            examplePath2Client.getExamplePath2PathParam("304", 10)
+        }
+
+        assertThat(result.message).isEqualTo("[304]: ")
```

Comment:
@cjbooms This looks dodgy. There seems to be no handling for the 3xx cases.
Messages are either `successful -> 200..299`, `bad request -> 400..499` or `server error -> 500..599`. 
Everything 3xx falls through into the else case in `http-util.kt.hbs` line 46.

Reply:
I don't think anyone at Zalando have used the API client for anything other than tests, so probably proper 3X handling never came up.